### PR TITLE
Misc async fixes

### PR DESCRIPTION
**Pull request by xconverge**

### CODE REVIEW

File: src/util/vscode-context.ts [hunk 1]
```diff
@@ -12,7 +12,7 @@ class VsCodeContextImpl {
     const prev = this.Get(key);
     if (!prev || prev !== value) {
       this.contextMap[key] = value;
-      return vscode.commands.executeCommand('setContext', key, value);
+      return await vscode.commands.executeCommand('setContext', key, value);
```
[the remark this answers is not in the document]
Comment (xconverge) [reply src/util/vscode-context.ts]: Answered here: https://github.com/eslint/eslint/blob/master/docs/rules/no-return-await.md

[the remark this answers is not in the document]
Comment (jpoon) [reply src/util/vscode-context.ts]: We want to return the promise.

**Comment by jpoon** [comment]
Wowza, I didn't realize there were so many spots we were unnecessarily awaiting :(

**Comment by jpoon** [comment]
@xconverge you mentioned on Slack you wanted me to confirm something?

File: extension.ts [hunk 2]
```diff
@@ -151,7 +151,7 @@ export async function activate(context: vscode.ExtensionContext) {
 
     // Delete modehandler once all tabs of this document have been closed
     for (let editorIdentity of ModeHandlerMap.getKeys()) {
-      let modeHandler = await ModeHandlerMap.get(editorIdentity);
+      const modeHandler = ModeHandlerMap.get(editorIdentity);
```
Comment (jpoon) [inline extension.ts]: We need `await` here. The resolved value is consumed immediately after.

File: extension.ts [hunk 3]
```diff
@@ -269,7 +269,7 @@ export async function activate(context: vscode.ExtensionContext) {
   registerCommand(context, 'vim.showQuickpickCmdLine', async () => {
     const modeHandler = await getAndUpdateModeHandler();
     await commandLine.PromptAndRun('', modeHandler.vimState);
-    modeHandler.updateView(modeHandler.vimState);
+    await modeHandler.updateView(modeHandler.vimState);
```
Comment (jpoon) [inline extension.ts]: I don't think we need to await here?

File: extension.ts [hunk 4]
```diff
@@ -308,7 +308,7 @@ export async function activate(context: vscode.ExtensionContext) {
   // Initialize mode handler for current active Text Editor at startup.
   if (vscode.window.activeTextEditor) {
     let mh = await getAndUpdateModeHandler();
-    mh.updateView(mh.vimState, { drawSelection: false, revealRange: false });
+    await mh.updateView(mh.vimState, { drawSelection: false, revealRange: false });
```
Comment (jpoon) [inline extension.ts]: Don't need to await this as the stuff thereafter is not reliant on an updated view.

File: src/actions/commands/actions.ts [hunk 5]
```diff
@@ -390,7 +390,7 @@ class CommandRecordMacro extends BaseCommand {
       // If register name is upper case, it means we are appending commands to existing register instead of overriding.
       let newRegister = new RecordedState();
       newRegister.registerName = register;
-      Register.putByKey(newRegister, register);
+      await Register.putByKey(newRegister, register);
```
Comment (jpoon) [inline src/actions/commands/actions.ts]: I don't think we need to await this?

File: src/actions/commands/actions.ts [hunk 6]
```diff
@@ -1648,7 +1648,7 @@ export class PutCommandVisual extends BaseCommand {
         true
       );
       const deletedRegister = await Register.getByKey(vimState.recordedState.registerName);
-      Register.putByKey(
+      await Register.putByKey(
```
Comment (jpoon) [inline src/actions/commands/actions.ts]: I don't think we need to await this?

File: src/actions/commands/actions.ts [hunk 7]
```diff
@@ -1660,7 +1660,7 @@ export class PutCommandVisual extends BaseCommand {
       await deleteResult.setCurrentMode(oldMode);
       deleteResult = await new PutCommand().exec(start, deleteResult, true);
       await deleteResult.setCurrentMode(resultMode);
-      Register.putByKey(
+      await Register.putByKey(
```
Comment (jpoon) [inline src/actions/commands/actions.ts]: I don't think we need to await this?

File: src/actions/operator.ts [hunk 8]
```diff
@@ -644,7 +644,7 @@ export class YankVisualBlockMode extends BaseOperator {
       toCopy += line + '\n';
     }
 
-    Register.put(toCopy, vimState, this.multicursorIndex);
+    await Register.put(toCopy, vimState, this.multicursorIndex);
```
Comment (jpoon) [inline src/actions/operator.ts]: I don't think we need to await this?

File: src/mode/modeHandler.ts [hunk 9]
```diff
@@ -587,7 +587,7 @@ export class ModeHandler implements vscode.Disposable {
       vimState.globalState.previousFullAction = vimState.recordedState;
 
       if (recordedState.isInsertion) {
-        Register.putByKey(recordedState, '.');
+        await Register.putByKey(recordedState, '.');
```
Comment (jpoon) [inline src/mode/modeHandler.ts]: I don't think we need to await this?

File: src/mode/modeHandler.ts [hunk 10]
```diff
@@ -959,7 +959,7 @@ export class ModeHandler implements vscode.Disposable {
           let cmd = await commandLine.ShowHistory(vimState.currentCommandlineText, this.vimState);
           if (cmd && cmd.length !== 0) {
             await commandLine.Run(cmd, this.vimState);
-            this.updateView(this.vimState);
+            await this.updateView(this.vimState);
```
Comment (jpoon) [inline src/mode/modeHandler.ts]: I don't think we need to await this?

File: src/neovim/neovim.ts [hunk 11]
```diff
@@ -152,7 +152,7 @@ export class Neovim implements vscode.Disposable {
     };
     vimState.currentRegisterMode =
       vimRegToVsReg[(await this.nvim.callFunction('getregtype', ['"'])) as string];
-    Register.put((await this.nvim.callFunction('getreg', ['"'])) as string, vimState);
+    await Register.put((await this.nvim.callFunction('getreg', ['"'])) as string, vimState);
```
Comment (jpoon) [inline src/neovim/neovim.ts]: I don't think we need to await this?

File: src/register/register.ts [hunk 12]
```diff
@@ -89,13 +93,18 @@ export class Register {
       if (Register.isValidUppercaseRegister(register)) {
         Register.appendMulticursorRegister(content, register, vimState, multicursorIndex as number);
       } else {
-        Register.putMulticursorRegister(content, register, vimState, multicursorIndex as number);
+        await Register.putMulticursorRegister(
```
Comment (jpoon) [inline src/register/register.ts]: I don't think we need to await this?

File: src/register/register.ts [hunk 13]
```diff
@@ -89,13 +93,18 @@ export class Register {
...
+          vimState,
+          multicursorIndex as number
+        );
       }
     } else {
       if (Register.isValidUppercaseRegister(register)) {
         Register.appendNormalRegister(content, register, vimState);
       } else {
-        Register.putNormalRegister(content, register, vimState);
+        await Register.putNormalRegister(content, register, vimState);
```
Comment (jpoon) [inline src/register/register.ts]: I don't think we need to await this?

File: src/register/register.ts [hunk 14]
```diff
@@ -153,7 +162,7 @@ export class Register {
         }
         clipboardText = clipboardText.replace(/\n$/, '');
 
-        Clipboard.Copy(clipboardText);
+        await Clipboard.Copy(clipboardText);
```
Comment (jpoon) [inline src/register/register.ts]: I don't think we need to await this?

File: src/register/register.ts [hunk 15]
```diff
@@ -214,13 +223,13 @@ export class Register {
...
-  private static putNormalRegister(
+  private static async putNormalRegister(
     content: RegisterContent,
     register: string,
     vimState: VimState
-  ): void {
+  ): Promise<void> {
     if (Register.isClipboardRegister(register)) {
-      Clipboard.Copy(content.toString());
+      await Clipboard.Copy(content.toString());
```
Comment (jpoon) [inline src/register/register.ts]: I don't think we need to await this?

File: src/register/register.ts [hunk 16]
```diff
@@ -273,17 +282,17 @@ export class Register {
...
     registerMode = RegisterMode.AscertainFromCurrentMode
-  ): void {
+  ): Promise<void> {
     if (!Register.isValidRegister(register)) {
       throw new Error(`Invalid register ${register}`);
     }
 
     if (Register.isClipboardRegister(register)) {
-      Clipboard.Copy(content.toString());
+      await Clipboard.Copy(content.toString());
```
Comment (jpoon) [inline src/register/register.ts]: I don't think we need to await this?

File: src/register/register.ts (hunk 15 above)
Comment (xconverge) [reply src/register/register.ts]: Why not? How do we make sure what is in the clipboard is what we intend to be in the clipboard then?

File: extension.ts (hunk 2 above)
Comment (xconverge) [reply extension.ts]: ModeHandlerMap.get() is not even async. Only getOrCreate() is async

**Comment by xconverge** [comment]
> 
> 
> @xconverge you mentioned on Slack you wanted me to confirm something?

taskQueue.ts line 87, this.runTasks does not have an await and scares me

**Comment by jpoon** [comment]
It looks fine to me?

I don't think it's necessary to wait for the promise on the `enqueue` there: https://github.com/VSCodeVim/Vim/blob/master/src/taskQueue.ts#L87

The individual tasks in the queue are awaited so they should be executed in order. https://github.com/VSCodeVim/Vim/blob/master/src/taskQueue.ts#L31

**Pull request closed without merging**